### PR TITLE
Remove master variants

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Removes "master" variants from the codebase and the datamodel :boom: 

no-masters

#### What should we test?

Existing products and product images should be unchanged. Creating and editing products should work correctly.
Including:
  + creating, editing products. 
    * Variant errors should now be better displayed (eg when 0 unit value)
    * adding/updating images
  + creating,editing variants. 
    * Note that new variants may no longer have default values like they used to.
    * adding/updating images
  + check products displayed in the shop the same as before

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: app/models/spree/variant.rb
Comment (mkllnk) [inline app/models/spree/variant.rb:1]: Shall we prepare for the final migration and ignore the is_master column?
```rb
self.ignored_columns = [:is_master]
```

Then we can remove it in a future PR without downtime concerns. It also validates on this PR that everything will still work. The migration spec will need to be removed again, or altered for this PR only.

**Review by mkllnk — APPROVED** [review]
Excellent pull request, very easy to follow the commits. And each commit makes the code simpler. This will free up some mental capacity working on OFN.

I have a tiny suggestion but it can follow in another PR as well.

P.S.: Great PR banner. :smile:

File: app/controllers/spree/admin/products_controller.rb
Comment (Matt-Yorkley) [reply app/controllers/spree/admin/products_controller.rb:232]: Aha! Yeah that's simpler :+1:

Comment (Matt-Yorkley) [reply app/controllers/spree/admin/products_controller.rb:232]: Hmmm now that I look at it again, I think this might just be dead code...

**Comment by jibees** [comment]
I just want to approve the PR simply because of the image used to illustrate it 🖤

File: app/controllers/spree/admin/products_controller.rb
Comment (Matt-Yorkley) [reply app/controllers/spree/admin/products_controller.rb]: I added a commit to delete it :ok_hand:

File: app/models/spree/variant.rb
Comment (Matt-Yorkley) [reply app/models/spree/variant.rb:1]: I just gave it a try in a separate branch here: https://github.com/Matt-Yorkley/openfoodnetwork/actions/runs/5200172972/jobs/9378552136

All the tests passed except those two in the `migrate_images_to_product_spec`.

So I guess we can either; remove those migration tests and disable the column, or rewrite the migration tests, or remove the column in a different PR, or... I guess we could just remove the column in this PR, as long as it runs after the data migration..? What do you think?

Comment (Matt-Yorkley) [reply app/models/spree/variant.rb:1]: Okay, I've added some commits to remove the column completely and remove the tests on the images data migration :ok_hand:

Comment (mkllnk) [reply app/models/spree/variant.rb:1]: >  remove the column completely

This will blow up during deploy because we run the migration and delete the column while the old code is still loaded and tries to use it. That's where ignoring the column first is handy to avoid downtime. This would be more important if we actually had zero-downtime deployments but I think that it's still worth to avoid errors. Do you agree?

File: db/migrate/20230607131645_remove_is_master_column.rb
Comment (mkllnk) [inline db/migrate/20230607131645_remove_is_master_column.rb:4]: I think that this will raise lots of errors during deploy.

**Review by mkllnk — CHANGES_REQUESTED** [review]
Good call on removing the spec again. It did it's job to test the migration but since migrations are used only once, we don't need to keep the test.

I'm just concerned about removing the column already.

File: db/migrate/20230607131645_remove_is_master_column.rb
Comment (Matt-Yorkley) [reply db/migrate/20230607131645_remove_is_master_column.rb]: Alright, I switched back to disabling the column in the class :ok_hand:

**Review by mkllnk — APPROVED** [review]
I just let Rubocop fix a little style issue. Not really relevant but :shrug: .

File: db/migrate/20230605133804_remove_master_variants.rb
Comment (dacook) [inline db/migrate/20230605133804_remove_master_variants.rb:3]: I was curious about this one, I'm guessing you have another branch in play that overlaps here?

**Review by dacook — COMMENTED** [review]
Fantastic work, this is an epic change! In fact, I thought some of these steps deserved to be their own separate PR (eg move some attributes, move images, remove master variant). But you've wrangled it all together well.

As mentioned, this removes a lot of bits of logic which makes the app easier to understand. Also I appreciate the cleanups along the way.

Firstly, I noticed a few changes that seemed unexpected to me. Can you please review my comments in:
1. [product_duplicator.rb](https://github.com/openfoodfoundation/openfoodnetwork/pull/10939/files/620a7fe4cc93928f96c4c2ba79993dc216010c05..de6c1de2af99f12b891f8237361fab3004c53921#diff-036c04457c974c5e5042eb0be71f36917b06a13277493a6c4ede14d7e1f8269b)
2. [scope_variants_to_search_spec.rb](https://github.com/openfoodfoundation/openfoodnetwork/pull/10939/files/620a7fe4cc93928f96c4c2ba79993dc216010c05..de6c1de2af99f12b891f8237361fab3004c53921#diff-5af409bc212eb446a4100e0d3e6029aecaaff06fd3e2e4b774449fb85fcdd94b)
3. [darkswarm_caching_spec.rb](https://github.com/openfoodfoundation/openfoodnetwork/pull/10939/files/80e55cae9b0cc727377589eaeaa349d01f60416a..5de039b2ca7c17e9a2e1e9d5d0273ca04b4dc45f#diff-3df65ba762ad165f1d7b8aa8f5e6f53673707414c54a745b194892f866607392)

Secondly, it looks like there are some changes to the admin editing experience. I'll add some notes to the testing section. Can you confirm:
* It looks like when you create a new variant, some fields were pre-populated before (`price, display_as, unit_value, unit_description`), and now they won't be (because we no longer have a 'default'). 

This seems fine to me, but maybe worth pointing out in the release notes.

Thirdly I thought it would be sensible to delay the actual deletion from the database for a later release. Although this PR is really solid, it's such a big change that wouldn't hurt to be cautious around.

And finally, just recording another thought for future improvement: seeing the prevalence of `variant: product.variants.first` in specs made me thing that we could enhance some of our factories to automatically pull this from a product. This would help reduce repetition in specs (eg `create(:order, product: product)`).

File: db/migrate/20230605133804_remove_master_variants.rb
Comment (mkllnk) [reply db/migrate/20230605133804_remove_master_variants.rb:1]: It's the main point of the pull request. Taking this out would mean taking a lot of other changes out. We would still need to filter by `is_master: false` by default, for example.

In general, yes, I would have split this PR into smaller chunks and this would have come later. But this PR was built as a whole and taking things out could be a lot more work, including testing that Matt has already done on the whole thing.

So I'm okay with this. It's slightly risky but I think that we can handle it.
Hm, this may actually cause some downtime as well. Maybe we need a rollout plan for this PR.

Comment (dacook) [reply db/migrate/20230605133804_remove_master_variants.rb:1]: > Taking this out would mean taking a lot of other changes out. We would still need to filter by `is_master: false` by default, for example.

Oh yeah, good point. I have to admit my brain was pretty fried by this point ;)

> Hm, this may actually cause some downtime as well. Maybe we need a rollout plan for this PR.

Yeah, if records are deleted before Puma (and Sidekiq) reload the new code, we will probably have some errors. Maybe not many of much consequence.. but would need to think through that before making that judgement.

If it can be deployed at a quiet time for each instance's timezone, it might be fine. 
Or otherwise we can schedule a downtime window. Hmm.. deployments can take quite some time (I haven't timed lately but it could be up to 20 mins). That's not great. Perhaps we can update the deploy playbook with a downtime option that only brings it down for the minimum time needed.

So maybe we should test out a live deployment and hope for the best?

**Comment by dacook** [comment]
I wondered if any of these changes require updates to documentation.

I noticed some changes in the serialisers, but I don't think any of those are visible, so we don't need to update the [API handbook](https://ofn-user-guide.gitbook.io/ofn-api-handbook/ofn-api-v0-unsupported/products), or [swagger API docs](https://app.swaggerhub.com/apis/Open-Food-Network/Open-Food-Network-API/0.1#/).

I don't _think_ that anything worth documenting has changed for the products admin interface, so we probably don't need to update the [products section](https://guide.openfoodnetwork.org/basic-features/products-1) of the user guide. 

Make sense, or could there be something else?

**Review by rioug — COMMENTED** [review]
Nice to see the code base simplified ! Epic :boom: 
I won't claim to have understood all the changes that have been made, but from coding point of view it looks good. 
I just spotted a small issue, looks like an oversight, see my comment.

File: spec/system/admin/reports_spec.rb
Comment (mkllnk) [reply spec/system/admin/reports_spec.rb:756]: Another confusing thing: Products and Variants both have an SKU. Before the Product SKU was stored on the master variant though. Now it's explicitly on the Product model.

Still, not sure why this update was needed. The product should still have an SKU, just stored differently.

File: spec/system/consumer/caching/darkswarm_caching_spec.rb
Comment (Matt-Yorkley) [reply spec/system/consumer/caching/darkswarm_caching_spec.rb:57]: The flakyness was caused by the timestamp not using millisecond precision and the test running, changing and checking the values within in a timespan of less than a second. It doesn't need the sleep at all after the changes here, it's always accurate :+1:

File: db/migrate/20230605133804_remove_master_variants.rb
Comment (Matt-Yorkley) [reply db/migrate/20230605133804_remove_master_variants.rb:3]: Yeah, that table isn't in use any more (`OptionValues` are gone now) and is in the process of being removed, but I wasn't sure if it would still be present or not at the exact time this PR was merged/deployed. So this migration will work correctly either way.

The point of interacting with that table here is that there's a foreign key constraint at the database level, so you can't delete a variant if there are associated records in the `spree_option_values_variants` table that match it's `id` (they would need to be deleted first). But if the table has been dropped already then it can be skipped.

File: spec/system/admin/reports_spec.rb
Comment (Matt-Yorkley) [reply spec/system/admin/reports_spec.rb:756]: It's because the test setup here was previously using `product.master` as the _variant_ that was being assigned to the line items being tested. I updated that in a previous commit, I think. 

Basically I found there were a lot of old tests where the test setup was essentially wrong as they were using the master variant as a kind of shortcut when setting up an order object with line items, but essentially those tests should have been updated during the big Spree upgrade (~2.5 years ago?) because master variants do not get associated to line items under any circumstances, and haven't done for a long time.

Does that make sense?

File: spec/models/spree/product_spec.rb
Comment (Matt-Yorkley) [reply spec/models/spree/product_spec.rb:19]: Hmmm yeah I think you're right. Previously it was duplicating the master variant with the `#duplicate_variant` method, which explicitly unsets the `sku`. I'll add a commit for that, it needs a slight code change and not just a change to the test :ok_hand:

Comment (Matt-Yorkley) [reply spec/models/spree/product_spec.rb:19]: Nice catch! Update here: https://github.com/openfoodfoundation/openfoodnetwork/pull/10939/commits/1d43a51d7322bf3966ae2b7ac126399a379e4de4

File: lib/spree/core/product_duplicator.rb
Comment (Matt-Yorkley) [reply lib/spree/core/product_duplicator.rb:30]: Hmmm I can't remember the exact details of why I cut this, but... `product.image` here will either be `nil` in which case we can completely ignore it, or it'll be a `Spree::Image` object in which case we should probably clone that object as-is and the additional check of `attached?` will just make an unnecessary additional query or two from the `Spree::Image` to the attachment?

File: spec/lib/open_food_network/scope_variants_to_search_spec.rb
Comment (Matt-Yorkley) [reply spec/lib/open_food_network/scope_variants_to_search_spec.rb]: This was an odd one because the actual code in the variant search was only really looking at variant SKUs and not product SKUs and that seemed reasonable. I took another look and I've updated it to match on both of them now here: https://github.com/openfoodfoundation/openfoodnetwork/pull/10939/commits/cda903c51168f8c10a9d67ab0864ac9a32fe412e

File: spec/system/admin/reports_spec.rb
Comment (rioug) [reply spec/system/admin/reports_spec.rb:756]: I think it does, Thanks !

File: spec/system/consumer/caching/darkswarm_caching_spec.rb
Comment (dacook) [reply spec/system/consumer/caching/darkswarm_caching_spec.rb:57]: Ah, thanks for explaining. So if the spec ran within 0.00001 of a second then it would fail? But that seems extremely unlikely :)

File: spec/lib/open_food_network/scope_variants_to_search_spec.rb
Comment (dacook) [reply spec/lib/open_food_network/scope_variants_to_search_spec.rb]: Bravo!

File: db/migrate/20230605133804_remove_master_variants.rb
Comment (dacook) [reply db/migrate/20230605133804_remove_master_variants.rb:1]: > Luckily we already have a playbook which does exactly that

Great, I didn't realise that! That will result in downtime for the entire deploy process. I haven't reviewed the deploy process, but was wondering if we can minimise the downtime by limiting it only to the core part of the process. 
I've created an issue, but I don't think we necessary need to do it this time.

**Review by dacook — APPROVED** [review]
Thanks for addressing those, I think this looks good now!

But I'm still unclear about the rollout plan. 

1.  Is maintenance mode/downtime necessary? If no errors of any consequence (eg corrupt data) are expected, then should we not use maintenance mode, and allow the app to be mostly functional, accepting that a few people might have temporary issues.
2. Do we need a rollback plan? (that is, be prepared to restore from backups if necessary)
3. Anything else we need to plan regarding the rollout?

**Comment by Matt-Yorkley** [comment]
I don't think we need scheduled downtime. A rollback plan is interesting, there's definitely some destructive changes here that are non-reversible. Maybe we could test out performing both the deployment of this PR and a rollback in staging and see how both of them look?

**Comment by dacook** [comment]
Maikel and I just discussed this and I'll summarise:

As Matt says, there's definitely some destructive changes here that are non-reversible. If any issues are identified, we may need to act quickly to fix them in place, and move forward. We would avoid a restore from backup at all costs, because it could result in lost orders, and we have no process for resolving that. 

I think it's still a good idea to test the restore from backup though, so we can know that it's an option. We perform a database backup as part of every deployment, but never use them, so the process is essentially untested. 

We thought that it would be worth deploying first for one instance and waiting a day to ensure there are no issues presented. We think AU would be a good choice because we can both provide support at the same time if needed.
Once that deployment is confirmed, it should be fine to go ahead and deploy all other instances, ideally during the day in AU, while most other regions are sleeping. This way we don't need to worry about downtime or any extra comms.

I think the cleanest way to organise it is to release this PR in its own separate release, so that it doesn't interrupt the usual process. After the usual release is deployed, we would deploy this one (first AU, then other instances) and should be finished before the end of the week.

**Comment by dacook** [comment]
Next steps: it looks like we have a broken spec that needs fixing.

Then I think we can move to testing, but hold off on merging until the next release is drafted. Then we merge and draft a second release.

At some point before release, we can also test deployment and rollback on a staging server.

**Comment by drummer83** [comment]
@Matt-Yorkley There's a linter error here as well.

**Review by mkllnk — APPROVED** [review]
I fixed the style complaint.

**Comment by sigmundpetersen** [comment]
There's a linter error 👇

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley,

I've staged and ran some tests - it's looking very good

- placing an order with existing products
- creating products -> error 500 if an image was set on the create product page; I'm not entirely sure this relates to this PR, as we're having issues around images (these should relate to aws3, which is being used in staging-UK)
- adding to OC and placing orders with the new products; checking stock changes

All as before, except that error 500 mentioned above:
https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/648f7bd066d5df0008504845?event_id=648f7bd000be8725b5410000&i=sk&m=nw

I'll run some more tests once we've fixed the linter error (and use a aws3 server for that).

**Comment by Matt-Yorkley** [comment]
I'll take a look at that Bugsnag error, it looks like it's something small that needs to be fixed before we can merge :+1:

**Comment by Matt-Yorkley** [comment]
@filipefurtad0 that image upload issue on the new product form should be fixed now :ok_hand:

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley - sorry still some linter issues... Maybe these come from the ruby upgrade?

**Comment by Matt-Yorkley** [comment]
> still some linter issues... Maybe these come from the ruby upgrade?

There's a lot of new ones from the Ruby upgrade, yeah. I think we can just ignore them here and test/merge this PR anyway.

**Comment by filipefurtad0** [comment]
> There's a lot of new ones from the Ruby upgrade, yeah. I think we can just ignore them here and test/merge this PR anyway.

Ok, cool - re-tested it now :checkered_flag:

1) tested products creation, with an image (previously broken) - now fine :green_circle: 
2) tested inventory, namely, variant overrides -> all as before
3) tested product import -> creating products from through import works as before
4) tested Products & Inventory reports -> all as before
5) products page: CRUD operations on variants -> all good
6) In fact, I've noticed no difference at all in the app, so I had the urge to see some DB changes:

```
openfoodnetwork=> select count(*) from spree_variants where is_master = true;
 count 
-------
     0
(1 row)

```

Pretty crazy...

I'll have a quick deploy on a staging server with AWS3, some quick tests around pics and merge.

PS:
One question - outside the testing scope: do we actually still need the `is_master` attribute, for `spree_variants`? Seems redundant now... But probably is used somewhere...

**Comment by Matt-Yorkley** [comment]
> do we actually still need the is_master attribute, for spree_variants? Seems redundant now (in the sense that it is always f), but probably is used somewhere/everywhere...

That was the subject of some debate during code review! Some of those last little bits will be temporarily left in, and removed in a subsequent PR after the rollout is done :+1:

**Comment by filipefurtad0** [comment]
Ok, we're exchanging on slack but keeping the PR here as a point of truth:

Deploying to staging-AU first triggered:
https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/64931f91894b520007d9ad1b?event_id=64931f9000becd4bc90e0000&i=sk&m=nw

After https://github.com/openfoodfoundation/openfoodnetwork/pull/10939/commits/6f5d3ceacc24e16b0f82d0a1bc20ba8777096087, deploying resulted in:

https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/6493243c894b520007d9cbc1?event_id=6493243c00becd51f28f0000&i=sk&m=nw

**Comment by filipefurtad0** [comment]
Alright, deployment worked fine in staging-AU; I've made some checks around images, checkout as well.

Looks good. Moving to ready to go!
Not sure we need code review on those last commits?

File: db/migrate/20230605133804_remove_master_variants.rb
Comment (mkllnk) [inline db/migrate/20230605133804_remove_master_variants.rb:23]: I agree that these shouldn't be there but deleting line items sounds dangerous. We should check the impact:

```
$ sql='SELECT count(*) FROM spree_line_items, spree_variants WHERE spree_variants.is_master = true AND spree_variants.id = spree_line_items.variant_id;'
$ ansible all-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.au | CHANGED | rc=0 >>
 count 
-------
 11027
(1 row)
openfoodnetwork.org.nz | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.net | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.ca | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.org.uk | CHANGED | rc=0 >>
 count 
-------
  2229
(1 row)
app.katuma.org | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
coopcircuits.fr | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.ie | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
csicsoka.org | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.de | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
openfoodnetwork.be | CHANGED | rc=0 >>
 count 
-------
     0
(1 row)
```

Only the two oldest instances have these line items. But there are many.

Comment (mkllnk) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: Is this data ancient or do we need to worry about this?
```
$ sql='SELECT max(spree_line_items.updated_at) FROM spree_line_items, spree_variants WHERE spree_variants.is_master = true AND spree_variants.id = spree_line_items.variant_id;'
$ ansible all-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.au | CHANGED | rc=0 >>
            max             
----------------------------
 2015-06-11 08:01:57.836125
(1 row)


openfoodnetwork.org.uk | CHANGED | rc=0 >>
            max             
----------------------------
 2020-08-04 01:55:15.996018
```

2015 sounds about right but UK has such line items updated in 2020?

Comment (mkllnk) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: I guess, if we want to play it safe, we could create a new pull request to fix this first. It could identify all these line items and point them to the first variant of the product instead of the master variant. That's maybe not 100% correct but it would keep the line items with quantity and price so that the order total makes sense.

What do you think?

**Review by mkllnk — COMMENTED** [review]
Great work but I have a concern about the latest addition.

File: db/migrate/20230605133804_remove_master_variants.rb
Comment (dacook) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: I was curious _how many_ orders there were in 2020, so I modified your query (thanks for sharing). It turns out there's just one order (at least, there's only line items created at one instant).

```
$ sql='SELECT DISTINCT spree_line_items.updated_at FROM spree_line_items, spree_variants WHERE spree_variants.is_master = true AND spree_variants.id = spree_line_items.variant_id ORDER BY spree_line_items.updated_at DESC LIMIT 20;'
$ ansible uk-prod -u openfoodnetwork -a "psql -h localhost openfoodnetwork ofn_user -c '$sql'"

openfoodnetwork.org.uk | CHANGED | rc=0 >>
         updated_at
----------------------------
 2020-08-04 01:55:15.996018
 2016-03-15 20:05:51.634713
 2015-06-10 18:24:50.69657
...
```

Comment (dacook) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: There's also one table that refers to spree_line_items: `spree_option_values_line_items`, so we could potentially be orphaning those. I didn't bother querying for them though, because it probably won't matter for ancient records.

Comment (dacook) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: I'm not sure what to suggest. Maikel's idea of migrating makes sense and may well be perfectly appropriate, but it's hard to know without reviewing all the orders. What if the price of the first variant is different to the master variant? So at the risk of wrong data, why not just delete.

I don't think we need to keep records that old, but deleting some and not others leaves an incorrect picture of history. If someone wants to analyse past data, they'll not be getting an accurate picture. So part of me wants to suggest we communicate that we're going to delete ALL orders before those dates (for UK and AU where it's relevant).

Hmm.. dunno.. Maikel's suggestion is probably fine, and nobody will notice anyway.

Comment (mkllnk) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: > What if the price of the first variant is different to the master variant?

The line item itself has a price stored in case you update the price of a variant later. So at least the order would be reflected correctly. There may be some information pull from the variant that could differ but the most important information would still be there through the product and the first variant.

Comment (dacook) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: 🤦 Ha, good point. So then it's a low risk update, and leaves data in a relatively consistent state. Seems worth doing to me.

Comment (Matt-Yorkley) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: Hmmm it's tricky. This issue of line items that reference a master variant is a scenario that in theory shouldn't have been valid or even possible for at least 5-6 years, and these old bits of data in theory should have been cleaned up at the time those changes were made. But a couple of servers have some really old data that's not in a nice state.

I think if we go down this route of trying to switch out each of the variants in these line items to a different variant we might end up with a huge can of worms to deal with. I imagine that some of the old products from that era will only have a master variant and not have a non-master variant, because it's before the point where having a non-master variant was made mandatory. And I'm not sure we even had proper soft-deletion on variants back then..?

Comment (Matt-Yorkley) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: Yep, I just confirmed the hypothesis; lots of the earliest variants in the database on UK are master and have no corresponding  non-master siblings.

Comment (Matt-Yorkley) [reply db/migrate/20230605133804_remove_master_variants.rb:23]: Sooooo.... okay how about; we hunt down these ancient line items where the corresponding variant has `is_master: true` and we just flip the flag on it to `is_master: false` in those cases? The variants will be intact from the point of view of the line items and there's no loss of relevant information.

File: db/migrate/20230605133804_remove_master_variants.rb
Comment (mkllnk) [inline db/migrate/20230605133804_remove_master_variants.rb:19]: That's brilliant! If the variant is referenced then we should turn it into a real variant. :star:

**Review by mkllnk — APPROVED** [review]

**Review by dacook — APPROVED** [review]
A much better idea, and surprisingly simple change! 🎉

**Comment by dacook** [comment]
We discovered that on UK prod, there were 5 offending `exchange_variants` records blocking the migration. They were pointing to master variants, and PG prevented us from orphaning them due to a foreign key constraint.
So I've updated the migration to flip these variants to non-master.

This also caused us to review all other records that point to the `spree_variants` table. Most were ok, but I found that _all_ master variant records had an unnecessary associated `spree_stock_items` record. These were silently ignored and orphaned due to no foreign key.  
So I've also updated the migration to remove these first.

https://openfoodnetwork.slack.com/archives/GK2T38QPJ/p1687481721929599

**Review by mkllnk — APPROVED** [review]

**Pull request merged**
